File: python/src/pylakekeeper/generic_tables.py
"""The generic-tables API surface — the methods that replace hand-rolled requests."""

from __future__ import annotations

from collections.abc import Iterator, Mapping
from typing import Any
from urllib.parse import quote

from .formats import GenericTableFormat, normalize_format
from .models import (
    GenericTableIdentifier,
    ListGenericTablesResponse,
    LoadGenericTableResponse,
)
from .transport import Transport
from .url import NamespaceLike, encode_namespace

_VENDED_HEADER = {"X-Iceberg-Access-Delegation": "vended-credentials"}


class GenericTables:
    """CRUD for Lakekeeper generic (non-Iceberg) tables, scoped to one warehouse."""

    def __init__(
        self,
        transport: Transport,
        warehouse: str,
        storage_overrides: Mapping[str, str] | None = None,
    ) -> None:
        self._t = transport
        self._warehouse = warehouse
        self._storage_overrides = dict(storage_overrides) if storage_overrides else None

    def _apply_overrides(self, resp: LoadGenericTableResponse) -> LoadGenericTableResponse:
        """Fold the client's storage overrides into the response's config.

        ``config`` takes precedence over the credential entries when the two are merged,
        so an override placed here reaches every downstream shape — ``lance_storage_options``,
        ``fsspec_kwargs`` and ``objects()`` — without each of them knowing about it.
        """
        if self._storage_overrides:
            resp.config = {**(resp.config or {}), **self._storage_overrides}
        return resp

    def _collection_path(self, namespace: NamespaceLike) -> str:
        return (
            f"/lakekeeper/v1/{quote(self._warehouse, safe='')}"
            f"/namespaces/{encode_namespace(namespace)}/generic-tables"
        )

    def _table_path(self, namespace: NamespaceLike, name: str) -> str:
        return f"{self._collection_path(namespace)}/{quote(name, safe='')}"

    def create(
        self,
        namespace: NamespaceLike,
        name: str,
        *,
        format: str | GenericTableFormat,
        base_location: str | None = None,
        doc: str | None = None,
        properties: Mapping[str, str] | None = None,
    ) -> LoadGenericTableResponse:
        """Create a generic table. Returns the load response for the new table.

        ``format`` accepts a :class:`~pylakekeeper.GenericTableFormat` or any string
        matching the server's shape rule (e.g. ``"lance"``); it is validated client-side.
        """
        body: dict[str, Any] = {"name": name, "format": normalize_format(format)}
        if base_location is not None:
            body["base-location"] = base_location
        if doc is not None:
            body["doc"] = doc
        if properties:
            body["properties"] = dict(properties)
        resp = self._t.request("POST", self._collection_path(namespace), json=body)
        return self._apply_overrides(LoadGenericTableResponse.model_validate(resp.json()))

    def load(
        self,
        namespace: NamespaceLike,
        name: str,
        *,
        vended: bool = False,
    ) -> LoadGenericTableResponse:
        """Load a generic table. With ``vended=True``, request inline STS credentials."""
        headers = _VENDED_HEADER if vended else None
        resp = self._t.request("GET", self._table_path(namespace, name), headers=headers)
        return self._apply_overrides(LoadGenericTableResponse.model_validate(resp.json()))

    def list(
        self,
        namespace: NamespaceLike,
        *,
        page_size: int = 100,
    ) -> Iterator[GenericTableIdentifier]:
        """Iterate every table identifier in a namespace, following ``next-page-token``."""
        path = self._collection_path(namespace)
        page_token: str | None = None
        while True:
            params: dict[str, Any] = {"pageSize": page_size}
            if page_token:
                params["pageToken"] = page_token
            resp = self._t.request("GET", path, params=params)
            page = ListGenericTablesResponse.model_validate(resp.json())
            yield from page.identifiers
            page_token = page.next_page_token
            if not page_token:
                return

    def drop(self, namespace: NamespaceLike, name: str) -> None:
        """Drop a generic table."""
        self._t.request("DELETE", self._table_path(namespace, name))

File: java/core/src/main/java/io/lakekeeper/client/GenericTables.java
package io.lakekeeper.client;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.lakekeeper.client.exception.LakekeeperException;
import io.lakekeeper.client.model.GenericTableIdentifier;
import io.lakekeeper.client.model.ListGenericTablesResponse;
import io.lakekeeper.client.model.LoadGenericTableResponse;

import java.net.URLEncoder;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.NoSuchElementException;

/** CRUD for Lakekeeper generic (non-Iceberg) tables, scoped to one warehouse. */
public final class GenericTables {
    private static final ObjectMapper MAPPER = new ObjectMapper();
    private static final String VENDED_HEADER_NAME = "X-Iceberg-Access-Delegation";
    private static final String VENDED_HEADER_VALUE = "vended-credentials";

    private final HttpTransport transport;
    private final String warehouse;

    GenericTables(HttpTransport transport, String warehouse) {
        this.transport = transport;
        this.warehouse = warehouse;
    }

    /**
     * Create a generic table.
     *
     * @param namespace dot-separated string or list of levels
     * @param name      table name
     * @param format    a {@link GenericTableFormat} or any valid format string
     */
    public LoadGenericTableResponse create(
            String namespace, String name, String format) {
        return create(namespace, name, format, null, null, null);
    }

    public LoadGenericTableResponse create(
            String namespace, String name, GenericTableFormat format) {
        return create(namespace, name, GenericTableFormat.normalize(format), null, null, null);
    }

    public LoadGenericTableResponse create(
            String namespace,
            String name,
            String format,
            String baseLocation,
            String doc,
            Map<String, String> properties) {
        Map<String, Object> body = new HashMap<>();
        body.put("name", name);
        body.put("format", GenericTableFormat.normalize(format));
        if (baseLocation != null) body.put("base-location", baseLocation);
        if (doc != null) body.put("doc", doc);
        if (properties != null && !properties.isEmpty()) body.put("properties", properties);

        HttpResponse<String> resp = transport.request("POST", collectionPath(namespace), null, toJson(body), null);
        return parse(resp.body(), LoadGenericTableResponse.class);
    }

    /** Load a generic table. */
    public LoadGenericTableResponse load(String namespace, String name) {
        return load(namespace, name, false);
    }

    /** Load a generic table, optionally requesting vended storage credentials. */
    public LoadGenericTableResponse load(String namespace, String name, boolean vended) {
        Map<String, String> headers = vended
                ? Collections.singletonMap(VENDED_HEADER_NAME, VENDED_HEADER_VALUE)
                : null;
        HttpResponse<String> resp = transport.request("GET", tablePath(namespace, name), null, null, headers);
        return parse(resp.body(), LoadGenericTableResponse.class);
    }

    /**
     * List all tables in a namespace, following pagination automatically.
     *
     * @return a lazy iterator; each call to {@link Iterator#next()} may issue an HTTP request
     */
    public Iterable<GenericTableIdentifier> list(String namespace) {
        return list(namespace, 100);
    }

    public Iterable<GenericTableIdentifier> list(String namespace, int pageSize) {
        return () -> new PageIterator(namespace, pageSize);
    }

    /** Drop a generic table. */
    public void drop(String namespace, String name) {
        transport.request("DELETE", tablePath(namespace, name), null, null, null);
    }

    private String collectionPath(String namespace) {
        return "/lakekeeper/v1/" + encode(warehouse)
                + "/namespaces/" + NamespaceEncoder.encode(namespace)
                + "/generic-tables";
    }

    private String tablePath(String namespace, String name) {
        return collectionPath(namespace) + "/" + encode(name);
    }

    private static String encode(String s) {
        return URLEncoder.encode(s, StandardCharsets.UTF_8);
    }

    private static String toJson(Object obj) {
        try {
            return MAPPER.writeValueAsString(obj);
        } catch (JsonProcessingException e) {
            throw new LakekeeperException("failed to serialize request body: " + e.getMessage(), e);
        }
    }

    private static <T> T parse(String json, Class<T> type) {
        try {
            return MAPPER.readValue(json, type);
        } catch (JsonProcessingException e) {
            throw new LakekeeperException("failed to parse response: " + e.getMessage(), e);
        }
    }

    private final class PageIterator implements Iterator<GenericTableIdentifier> {
        private final String namespace;
        private final int pageSize;
        private List<GenericTableIdentifier> current = new ArrayList<>();
        private int index = 0;
        private String nextPageToken = null;
        private boolean firstPage = true;

        PageIterator(String namespace, int pageSize) {
            this.namespace = namespace;
            this.pageSize = pageSize;
        }

        @Override
        public boolean hasNext() {
            if (index < current.size()) return true;
            if (!firstPage && nextPageToken == null) return false;
            fetchPage();
            return index < current.size();
        }

        @Override
        public GenericTableIdentifier next() {
            if (!hasNext()) throw new NoSuchElementException();
            return current.get(index++);
        }

        private void fetchPage() {
            firstPage = false;
            Map<String, String> params = new HashMap<>();
            params.put("pageSize", String.valueOf(pageSize));
            if (nextPageToken != null) params.put("pageToken", nextPageToken);
            HttpResponse<String> resp = transport.request("GET", collectionPath(namespace), params, null, null);
            ListGenericTablesResponse page = parse(resp.body(), ListGenericTablesResponse.class);
            current = page.getIdentifiers();
            index = 0;
            nextPageToken = page.getNextPageToken();
        }
    }
}
